src/agent/src/device.rs

```diff
@@ -436,217 +426,242 @@ fn scan_scsi_bus(scsi_addr: &str) -> Result<()> {
+            specdev.minor = update.guest_minor;
+            if let Some(final_path) = update.final_path.take() {
+                specdev.path = final_path;
+            }
+
+            for (host, guest) in update.pci.drain(..) {
+                if let Some(other_guest) = pcimap.insert(host, guest) {
+                    return Err(anyhow!(
+                        "Conflicting guest address for host device {} ({} versus {})",
+                        host,
+                        guest,
+                        other_guest
+                    ));
+                }
+            }
+
+            if resupdates
+                .insert((&specdev.r#type, host_major, host_minor), update)
+                .is_some()
+            {
+                return Err(anyhow!(
+                    "Conflicting device information for host_major={} host_minor={}",
+                    host_major,
+                    host_minor
+                ));
+            }
+        }
     }
 
-    let meta = fs::metadata(vm_path)?;
-    let dev_id = meta.rdev();
-    unsafe {
-        major_id = major(dev_id);
-        minor_id = minor(dev_id);
+    // Make sure we applied all of our updates
+    if !updates.is_empty() {
+        return Err(anyhow!(
+            "Missing devices in OCI spec: {:?}",
+            updates.into_keys().collect::<Vec<&str>>()
+        ));
     }
 
-    info!(
-        sl!(),
-        "update_spec_device(): vm_path={}, major: {}, minor: {}\n", vm_path, major_id, minor_id
-    );
-
-    if let Some(idxdata) = devidx.0.get(host_path) {
-        let dev = &mut linux.devices[idxdata.idx];
-        let host_major = dev.major;
-        let host_minor = dev.minor;
-
-        dev.major = major_id as i64;
-        dev.minor = minor_id as i64;
-        dev.path = final_path.to_string();
-
-        info!(
-            sl!(),
-            "change the device from path: {} major: {} minor: {} to vm device path: {} major: {} minor: {}",
-            host_path,
-            host_major,
-            host_minor,
-            dev.path,
-            dev.major,
-            dev.minor,
-        );
+    if let Some(resources) = linux.resources.as_mut() {
+        for r in &mut resources.devices {
+            if let (Some(host_major), Some(host_minor)) = (r.major, r.minor) {
+                if let Some(update) = resupdates.get(&(r.r#type.as_str(), host_major, host_minor)) {
+                    info!(
+                        sl!(),
+                        "update_spec_devices() updating resource";
+                        "type" => &r.r#type,
+                        "host_major" => host_major,
+                        "host_minor" => host_minor,
+                        "guest_major" => update.guest_major,
+                        "guest_minor" => update.guest_minor,
+                    );
+
+                    r.major = Some(update.guest_major);
+                    r.minor = Some(update.guest_minor);
+                }
+            }
+        }
+    }
```

Comment:
This isn't a huge block but could it be replaced by a call to something like `apply_resource_updates()?` ?

Reply:
So, it turns out that splitting off the resource updates causes a real lifetime mess.  With the update moved to a separate function, the compiler is not longer able to see that we don't change `specdev.r#type`, which means the immutable borrow of `specdev.r#type` we need for the hash lookup conflicts with the mutable borrow of `r` itself.  I don't think shortening this not-huge function is worth the trouble if disentangling this.